File: scripts/image_processing/layout_io.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any


REQUIRED_MARKER_ROLES = (
    "top-left",
    "top-right",
    "middle-left",
    "middle-right",
    "bottom-left",
    "bottom-right",
)


@dataclass(frozen=True)
class Rect:
    x: float
    y: float
    width: float
    height: float

    @property
    def center(self) -> tuple[float, float]:
        return (self.x + self.width / 2, self.y + self.height / 2)


@dataclass(frozen=True)
class LayoutMarker:
    role: str
    rect: Rect

    @property
    def center_mm(self) -> tuple[float, float]:
        return self.rect.center


@dataclass(frozen=True)
class ObjectiveOption:
    block_id: str
    question_number: int
    label: str
    rect: Rect


@dataclass(frozen=True)
class LayoutPage:
    card_id: str
    page_number: int
    width_mm: float
    height_mm: float
    markers: dict[str, LayoutMarker]
    objective_options: tuple[ObjectiveOption, ...] = ()


def _rect_from_json(value: dict[str, Any]) -> Rect:
    try:
        return Rect(
            x=float(value["x"]),
            y=float(value["y"]),
            width=float(value["width"]),
            height=float(value["height"]),
        )
    except KeyError as error:
        raise ValueError(f"Rect missing field: {error}") from error


def _objective_options_from_page(page_data: dict[str, Any]) -> tuple[ObjectiveOption, ...]:
    options: list[ObjectiveOption] = []

    for block in page_data.get("blocks", []):
        if not isinstance(block, dict) or block.get("type") != "objective":
            continue

        block_id = str(block.get("blockId") or "")
        for item in block.get("items", []):
            if not isinstance(item, dict):
                continue
            question_number = int(item.get("questionNumber", 0))
            if question_number <= 0:
                continue
            for option in item.get("options", []):
                if not isinstance(option, dict):
                    continue
                label = str(option.get("label") or "")
                rect_data = option.get("rect")
                if not label or not isinstance(rect_data, dict):
                    continue
                options.append(
                    ObjectiveOption(
                        block_id=block_id,
                        question_number=question_number,
                        label=label,
                        rect=_rect_from_json(rect_data),
                    )
                )

    if options:
        return tuple(sorted(options, key=lambda item: (item.question_number, item.label)))

    for element in page_data.get("elements", []):
        if not isinstance(element, dict) or element.get("type") != "objective_option":
            continue
        label = str(element.get("option") or "")
        rect_data = element.get("rect")
        question_number = int(element.get("questionNumber", 0))
        if not label or question_number <= 0 or not isinstance(rect_data, dict):
            continue
        options.append(
            ObjectiveOption(
                block_id=str(element.get("blockId") or ""),
                question_number=question_number,
                label=label,
                rect=_rect_from_json(rect_data),
            )
        )

    return tuple(sorted(options, key=lambda item: (item.question_number, item.label)))


def load_layout_page(layout_path: str | Path, page_number: int) -> LayoutPage:
    path = Path(layout_path)
    if not path.exists():
        raise FileNotFoundError(f"Layout JSON not found: {path}")

    with path.open("r", encoding="utf-8") as file:
        layout = json.load(file)

    card_id = str(layout.get("cardId") or "")
    if not card_id:
        raise ValueError(f"Layout JSON has no cardId: {path}")

    pages = layout.get("pages")
    if not isinstance(pages, list):
        raise ValueError(f"Layout JSON pages must be a list: {path}")

    page_data = next((page for page in pages if int(page.get("pageNumber", -1)) == page_number), None)
    if page_data is None:
        raise ValueError(f"Page {page_number} not found in layout JSON: {path}")

    marker_items = page_data.get("markers")
    if not isinstance(marker_items, list):
        raise ValueError(f"Page {page_number} has no marker list: {path}")

    markers: dict[str, LayoutMarker] = {}
    for marker in marker_items:
        role = str(marker.get("role") or "")
        if not role:
            continue
        markers[role] = LayoutMarker(role=role, rect=_rect_from_json(marker["rect"]))

    missing = [role for role in REQUIRED_MARKER_ROLES if role not in markers]
    if missing:
        raise ValueError(f"Page {page_number} is missing markers: {', '.join(missing)}")

    return LayoutPage(
        card_id=card_id,
        page_number=page_number,
        width_mm=float(page_data.get("width") or layout.get("width") or 210),
        height_mm=float(page_data.get("height") or layout.get("height") or 297),
        markers={role: markers[role] for role in REQUIRED_MARKER_ROLES},
        objective_options=_objective_options_from_page(page_data),
    )


def a4_pixel_size(width_mm: float, height_mm: float, dpi: int) -> tuple[int, int]:
    if dpi <= 0:
        raise ValueError("DPI must be positive")
    return (round(width_mm / 25.4 * dpi), round(height_mm / 25.4 * dpi))


def marker_centers_px(page: LayoutPage, dpi: int) -> dict[str, tuple[float, float]]:
    scale = dpi / 25.4
    return {
        role: (marker.center_mm[0] * scale, marker.center_mm[1] * scale)
        for role, marker in page.markers.items()
    }

File: native/AnswerCardRecognizer/answer-card-recognizer/layout_io.cpp
#include "layout_io.hpp"

#include "common.hpp"

#include <algorithm>
#include <cmath>
#include <fstream>
#include <stdexcept>

#include <nlohmann/json.hpp>

using json = nlohmann::json;

const std::vector<std::string> REQUIRED_MARKER_ROLES = {
    "top-left",
    "top-right",
    "middle-left",
    "middle-right",
    "bottom-left",
    "bottom-right",
};

std::pair<double, double> Rect::center() const {
    return {x + width / 2.0, y + height / 2.0};
}

std::pair<double, double> LayoutMarker::center_mm() const {
    return rect.center();
}

static Rect rect_from_json(const json& value) {
    if (!value.is_object()) {
        throw std::runtime_error("Rect must be an object");
    }
    for (const auto* key : {"x", "y", "width", "height"}) {
        if (!value.contains(key)) {
            throw std::runtime_error(std::string("Rect missing field: ") + key);
        }
    }
    return Rect{
        value.at("x").get<double>(),
        value.at("y").get<double>(),
        value.at("width").get<double>(),
        value.at("height").get<double>(),
    };
}

static std::string question_number_from_json(const json& value) {
    if (value.is_string()) {
        return value.get<std::string>();
    }
    if (value.is_number_integer()) {
        return std::to_string(value.get<int>());
    }
    if (value.is_number_float()) {
        return std::to_string(value.get<double>());
    }
    return "";
}

static std::vector<ObjectiveOption> objective_options_from_page(const json& page_data) {
    std::vector<ObjectiveOption> options;

    if (page_data.contains("blocks") && page_data.at("blocks").is_array()) {
        for (const auto& block : page_data.at("blocks")) {
            if (!block.is_object() || block.value("type", "") != "objective") {
                continue;
            }
            const std::string block_id = block.value("blockId", "");
            if (!block.contains("items") || !block.at("items").is_array()) {
                continue;
            }
            for (const auto& item : block.at("items")) {
                if (!item.is_object()) {
                    continue;
                }
                const int question_number = item.value("questionNumber", 0);
                if (question_number <= 0 || !item.contains("options") || !item.at("options").is_array()) {
                    continue;
                }
                for (const auto& option : item.at("options")) {
                    if (!option.is_object()) {
                        continue;
                    }
                    const std::string label = option.value("label", "");
                    if (label.empty() || !option.contains("rect") || !option.at("rect").is_object()) {
                        continue;
                    }
                    options.push_back(ObjectiveOption{block_id, question_number, label, rect_from_json(option.at("rect"))});
                }
            }
        }
    }

    if (!options.empty()) {
        std::sort(options.begin(), options.end(), [](const auto& left, const auto& right) {
            return std::tie(left.question_number, left.label) < std::tie(right.question_number, right.label);
        });
        return options;
    }

    if (page_data.contains("elements") && page_data.at("elements").is_array()) {
        for (const auto& element : page_data.at("elements")) {
            if (!element.is_object() || element.value("type", "") != "objective_option") {
                continue;
            }
            const std::string label = element.value("option", "");
            const int question_number = element.value("questionNumber", 0);
            if (label.empty() || question_number <= 0 || !element.contains("rect") || !element.at("rect").is_object()) {
                continue;
            }
            options.push_back(ObjectiveOption{
                element.value("blockId", ""),
                question_number,
                label,
                rect_from_json(element.at("rect")),
            });
        }
    }

    std::sort(options.begin(), options.end(), [](const auto& left, const auto& right) {
        return std::tie(left.question_number, left.label) < std::tie(right.question_number, right.label);
    });
    return options;
}

static std::vector<SubjectiveScoreCell> subjective_score_cells_from_page(const json& page_data) {
    std::vector<SubjectiveScoreCell> cells;

    if (page_data.contains("blocks") && page_data.at("blocks").is_array()) {
        for (const auto& block : page_data.at("blocks")) {
            if (!block.is_object() || block.value("type", "") != "subjective") {
                continue;
            }
            const std::string block_id = block.value("blockId", "");
            if (!block.contains("questions") || !block.at("questions").is_array()) {
                continue;
            }
            for (const auto& question : block.at("questions")) {
                if (!question.is_object() || !question.contains("scoreCells") || !question.at("scoreCells").is_array()) {
                    continue;
                }
                const std::string question_id = question.value("questionId", "");
                const std::string question_number = question.contains("questionNumber") ? question_number_from_json(question.at("questionNumber")) : "";
                const double max_score = question.value("score", 0.0);
                for (const auto& cell : question.at("scoreCells")) {
                    if (!cell.is_object() || !cell.contains("rect") || !cell.at("rect").is_object()) {
                        continue;
                    }
                    cells.push_back(SubjectiveScoreCell{
                        block_id,
                        question_id,
                        question_number,
                        cell.value("score", 0.0),
                        max_score,
                        rect_from_json(cell.at("rect")),
                    });
                }
            }
        }
    }

    if (!cells.empty()) {
        std::sort(cells.begin(), cells.end(), [](const auto& left, const auto& right) {
            return std::tie(left.block_id, left.question_id, left.score) < std::tie(right.block_id, right.question_id, right.score);
        });
        return cells;
    }

    if (page_data.contains("elements") && page_data.at("elements").is_array()) {
        for (const auto& element : page_data.at("elements")) {
            if (!element.is_object() || element.value("type", "") != "score_cell") {
                continue;
            }
            if (!element.contains("rect") || !element.at("rect").is_object()) {
                continue;
            }
            cells.push_back(SubjectiveScoreCell{
                element.value("blockId", ""),
                element.value("questionId", ""),
                element.contains("questionNumber") ? question_number_from_json(element.at("questionNumber")) : "",
                element.value("score", 0.0),
                0.0,
                rect_from_json(element.at("rect")),
            });
        }
    }

    std::sort(cells.begin(), cells.end(), [](const auto& left, const auto& right) {
        return std::tie(left.block_id, left.question_id, left.score) < std::tie(right.block_id, right.question_id, right.score);
    });
    return cells;
}

static std::vector<StudentDigit> student_digits_from_page(const json& page_data) {
    std::vector<StudentDigit> digits;
    if (!page_data.contains("elements") || !page_data.at("elements").is_array()) {
        return digits;
    }

    for (const auto& element : page_data.at("elements")) {
        if (!element.is_object() || element.value("type", "") != "student_digit") {
            continue;
        }
        if (!element.contains("rect") || !element.at("rect").is_object()) {
            continue;
        }
        digits.push_back(StudentDigit{
            element.value("digitIndex", 0),
            element.value("digit", 0),
            rect_from_json(element.at("rect")),
        });
    }

    std::sort(digits.begin(), digits.end(), [](const auto& left, const auto& right) {
        return std::tie(left.digit_index, left.digit) < std::tie(right.digit_index, right.digit);
    });
    return digits;
}

static std::vector<LayoutBlockCrop> block_crops_from_page(const json& page_data) {
    std::vector<LayoutBlockCrop> crops;
    if (!page_data.contains("blocks") || !page_data.at("blocks").is_array()) {
        return crops;
    }

    for (const auto& block : page_data.at("blocks")) {
        if (!block.is_object()) {
            continue;
        }
        const std::string block_id = block.value("blockId", "");
        const std::string block_type = block.value("type", "");
        if (block_id.empty() || block_type.empty()) {
            continue;
        }

        std::vector<std::string> question_numbers;
        if (block_type == "objective" && block.contains("items") && block.at("items").is_array()) {
            for (const auto& item : block.at("items")) {
                if (!item.is_object() || !item.contains("questionNumber")) {
                    continue;
                }
                const std::string number = question_number_from_json(item.at("questionNumber"));
                if (!number.empty()) {
                    question_numbers.push_back(number);
                }
            }
        } else if (block_type == "subjective" && block.contains("questions") && block.at("questions").is_array()) {
            for (const auto& question : block.at("questions")) {
                if (!question.is_object() || !question.contains("questionNumber")) {
                    continue;
                }
                const std::string number = question_number_from_json(question.at("questionNumber"));
                if (!number.empty()) {
                    question_numbers.push_back(number);
                }
            }
        }

        if (question_numbers.empty()) {
            continue;
        }
        const json* rect_value = nullptr;
        if (block.contains("frameRect") && block.at("frameRect").is_object()) {
            rect_value = &block.at("frameRect");
        } else if (block.contains("rect") && block.at("rect").is_object()) {
            rect_value = &block.at("rect");
        }
        if (!rect_value) {
            continue;
        }

        std::sort(question_numbers.begin(), question_numbers.end());
        question_numbers.erase(std::unique(question_numbers.begin(), question_numbers.end()), question_numbers.end());
        crops.push_back(LayoutBlockCrop{
            block_id,
            block.value("title", ""),
            block_type,
            rect_from_json(*rect_value),
            question_numbers,
        });
    }

    return crops;
}

LayoutPage load_layout_page(const std::filesystem::path& layout_path, int page_number) {
    if (!std::filesystem::exists(layout_path)) {
        throw std::runtime_error("Layout JSON not found: " + path_to_utf8(layout_path));
    }

    std::ifstream input(layout_path, std::ios::binary);
    if (!input) {
        throw std::runtime_error("Failed to open layout JSON: " + path_to_utf8(layout_path));
    }

    json layout;
    input >> layout;

    const std::string card_id = layout.value("cardId", "");
    if (card_id.empty()) {
        throw std::runtime_error("Layout JSON has no cardId: " + path_to_utf8(layout_path));
    }

    if (!layout.contains("pages") || !layout.at("pages").is_array()) {
        throw std::runtime_error("Layout JSON pages must be a list: " + path_to_utf8(layout_path));
    }

    const json* page_data = nullptr;
    for (const auto& page : layout.at("pages")) {
        if (page.is_object() && page.value("pageNumber", -1) == page_number) {
            page_data = &page;
            break;
        }
    }
    if (!page_data) {
        throw std::runtime_error("Page " + std::to_string(page_number) + " not found in layout JSON: " + path_to_utf8(layout_path));
    }
    if (!page_data->contains("markers") || !page_data->at("markers").is_array()) {
        throw std::runtime_error("Page " + std::to_string(page_number) + " has no marker list: " + path_to_utf8(layout_path));
    }

    std::map<std::string, LayoutMarker> markers;
    for (const auto& marker : page_data->at("markers")) {
        if (!marker.is_object()) {
            continue;
        }
        const std::string role = marker.value("role", "");
        if (role.empty() || !marker.contains("rect")) {
            continue;
        }
        markers[role] = LayoutMarker{role, rect_from_json(marker.at("rect"))};
    }

    std::vector<std::string> missing;
    for (const auto& role : REQUIRED_MARKER_ROLES) {
        if (!markers.contains(role)) {
            missing.push_back(role);
        }
    }
    if (!missing.empty()) {
        std::string message = "Page " + std::to_string(page_number) + " is missing markers: ";
        for (size_t index = 0; index < missing.size(); ++index) {
            if (index > 0) {
                message += ", ";
            }
            message += missing[index];
        }
        throw std::runtime_error(message);
    }

    std::map<std::string, LayoutMarker> required_markers;
    for (const auto& role : REQUIRED_MARKER_ROLES) {
        required_markers[role] = markers.at(role);
    }

    return LayoutPage{
        card_id,
        page_number,
        page_data->value("width", layout.value("width", 210.0)),
        page_data->value("height", layout.value("height", 297.0)),
        required_markers,
        objective_options_from_page(*page_data),
        student_digits_from_page(*page_data),
        subjective_score_cells_from_page(*page_data),
        block_crops_from_page(*page_data),
    };
}

std::pair<int, int> layout_pixel_size(double width_mm, double height_mm, int dpi) {
    if (dpi <= 0) {
        throw std::runtime_error("DPI must be positive");
    }
    return {
        static_cast<int>(std::llround(width_mm / 25.4 * dpi)),
        static_cast<int>(std::llround(height_mm / 25.4 * dpi)),
    };
}

std::map<std::string, std::pair<double, double>> marker_centers_px(const LayoutPage& page, int dpi) {
    const double scale = static_cast<double>(dpi) / 25.4;
    std::map<std::string, std::pair<double, double>> centers;
    for (const auto& [role, marker] : page.markers) {
        const auto [x, y] = marker.center_mm();
        centers[role] = {x * scale, y * scale};
    }
    return centers;
}
